tests/testutils/golden.go

```diff
@@ -128,8 +129,15 @@ func RunMetricsCollectionTest(t *testing.T, configFile string, expectedFilePath
 			assert.Fail(tt, "No metrics collected")
 			return
 		}
-		err := pmetrictest.CompareMetrics(expected, sink.AllMetrics()[len(sink.AllMetrics())-1],
-			opts.compareMetricsOptions...)
-		assert.NoError(tt, err)
+		var errs error
+		for _, m := range sink.AllMetrics() {
+			err := pmetrictest.CompareMetrics(expected, m,
+				opts.compareMetricsOptions...)
+			if err == nil {
+				return
+			}
+			errs = errors.Join(errs, err)
```

Comment:
I have a couple of concerns here:
1. Joined errors will be very hard to read.
2. We are scanning all of them every second which can increase time to run the tests

I'd suggest something like this:

```
        var checkedMetrics int
	assert.EventuallyWithT(t, func(tt *assert.CollectT) {
		if len(sink.AllMetrics()) == 0 {
			assert.Fail(tt, "No metrics collected")
			return
		}
		if len(sink.AllMetrics()) > checkedMetrics {
			checkedMetrics++
		}
		assert.NoError(tt, pmetrictest.CompareMetrics(expected, sink.AllMetrics()[checkedMetrics-1], opts.compareMetricsOptions...))
	}, 30*time.Second, 200*time.Millisecond)
```

I assume `assert.EventuallyWithT` cannot run concurrently here

Reply:
yes I am a bit lazy and we can incur a performance cost. Your method would work but we'd need to read all the new metrics, not just the new one, as multiple batches can be received. I can rework this code to accommodate.

Reply:
The method I posted read all the metrics

Reply:
On the other hand, this can slow down the validation by the provided frequency. So, you're right, it's better to read all of them as soon as they arrive